File: WebShop/WebShop-API/Services/TransactionService.cs
using WebShop_API.Database.Entities;
using WebShop_API.DTOs.Product;

namespace WebShop_API.Services
{
    public interface ITransactionService
    {
        Task<List<StaticTransactionResponse>> GetAllTransactions();
        Task<DirectTransactionResponse> GetByTransactionId(int transactionId);
        Task<DirectTransactionResponse> CreateTransaction(TransactionRequest request);
        Task<DirectTransactionResponse> UpdateTransaction(int transactionId, TransactionRequest request);
        Task<List<StaticOrderResponse>> GetAllOrders();
        Task<DirectOrderResponse> GetByOrderId(int orderId);
        Task<DirectOrderResponse> CreateOrder(OrderRequest request);
        Task<DirectOrderResponse> UpdateOrder(int orderId, OrderRequest request);

    }

    public class TransactionService : ITransactionService
    {

        private readonly ITransactionRepository m_transactionRepository;
        private readonly IOrderRepository m_orderRepository;
        private readonly IMapper m_mapper;
        private readonly IProductRepository m_productRepository;

        public TransactionService(ITransactionRepository transactionRepository, IOrderRepository orderRepository, IMapper mapper, IProductRepository productRepository)
        {
            m_transactionRepository = transactionRepository;
            m_orderRepository = orderRepository;
            m_mapper = mapper;
            m_productRepository = productRepository;
        }

        public async Task<DirectOrderResponse> CreateOrder(OrderRequest request)
        {
            Order order = await m_orderRepository.Create(m_mapper.Map<Order>(request));
            if(order != null)
            {
                return m_mapper.Map<DirectOrderResponse>(order);
            }

            return null;
        }

        public async Task<DirectTransactionResponse> CreateTransaction(TransactionRequest request)
        {
            Transaction transaction = await m_transactionRepository.Create(m_mapper.Map<Transaction>(request));
            if (transaction != null)
            {
                await m_productRepository.UpdateQuantity(transaction.ProductID, transaction.ProductAmount);
                return m_mapper.Map<DirectTransactionResponse>(transaction);
            }

            return null;
        }

        public async Task<List<StaticOrderResponse>> GetAllOrders()
        {
            List<Order> orders = await m_orderRepository.GetAll();
            if(orders != null)
            {
                return orders.Select(order => m_mapper.Map<StaticOrderResponse>(order)).ToList();
            }

            return null;
        }

        public async Task<List<StaticTransactionResponse>> GetAllTransactions()
        {
            List<Transaction> transactions = await m_transactionRepository.GetAll();
            if (transactions != null)
            {
                return transactions.Select(transaction => m_mapper.Map<StaticTransactionResponse>(transaction)).ToList();
            }

            return null;
        }

        public async Task<DirectOrderResponse> GetByOrderId(int orderId)
        {
            Order order = await m_orderRepository.GetById(orderId);
            if (order != null)
            {
                return m_mapper.Map<DirectOrderResponse>(order);
            }

            return null;
        }

        public async Task<DirectTransactionResponse> GetByTransactionId(int transactionId)
        {
            Transaction transaction = await m_transactionRepository.GetById(transactionId);
            if (transaction != null)
            {
                return m_mapper.Map<DirectTransactionResponse>(transaction);
            }

            return null;
        }

        public async Task<DirectOrderResponse> UpdateOrder(int orderId, OrderRequest request)
        {
            Order order = await m_orderRepository.Update(orderId, m_mapper.Map<Order>(request));

            if (order != null)
            {
                return m_mapper.Map<DirectOrderResponse>(order);
            }

            return null;
        }

        public async Task<DirectTransactionResponse> UpdateTransaction(int transactionId, TransactionRequest request)
        {
            Transaction transaction = await m_transactionRepository.Update(transactionId, m_mapper.Map<Transaction>(request));
            if (transaction != null)
            {
                return m_mapper.Map<DirectTransactionResponse>(transaction);
            }

            return null;
        }
    }
}

File: WebShop/WebShop-Client/src/app/_services/transaction.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from 'src/environments/environment';
import { StaticOrderResponse } from '../_models/order';
import { DirectOrderResponse } from '../_models/order/DirectOrderResponse';
import { OrderRequest } from '../_models/order/OrderRequest';
import { DirectTransactionResponse, StaticTransactionResponse, TransactionRequest } from '../_models/transaction';

@Injectable({
  providedIn: 'root'
})
export class TransactionService {

	// API URL
	private url: string = environment.ApiUrl + '/transaction/';
	private transactionUrl: string = this.url + 'transaction';
	private orderUrl: string = this.url + 'order';


  	constructor(private http: HttpClient) { }

	public GetAllTransaction(): Observable<StaticTransactionResponse[]> {
		return this.http.get<StaticTransactionResponse[]>(this.transactionUrl);
	}

	public GetTransaction(transactionId: number): Observable<DirectTransactionResponse> {
		return this.http.get<DirectTransactionResponse>(`${this.transactionUrl}/${transactionId}`);
	}

	public CreateTransaction(request: TransactionRequest): Observable<DirectTransactionResponse> {
		return this.http.post<DirectTransactionResponse>(this.transactionUrl, request);
	}

	public UpdateTransaction(transactionId: number, request: TransactionRequest): Observable<DirectTransactionResponse> {
		return this.http.put<DirectTransactionResponse>(`${this.transactionUrl}/${transactionId}`, request);
	}

	public GetAllOrders(): Observable<StaticOrderResponse[]> {
		return this.http.get<StaticOrderResponse[]>(this.orderUrl);
	}

	public GetOrder(orderId: number): Observable<DirectOrderResponse> {
		return this.http.get<DirectOrderResponse>(`${this.orderUrl}/${orderId}`);
	}

	public CreateOrder(request: OrderRequest): Observable<DirectOrderResponse> {
		return this.http.post<DirectOrderResponse>(this.orderUrl, request);
	}

	public UpdateOrder(orderId: number, request: OrderRequest): Observable<DirectOrderResponse> {
		return this.http.put<DirectOrderResponse>(`${this.orderUrl}/${orderId}`, request);
	}
}
